tiledb/cf/engines/xarray_engine.py

```diff
@@ -0,0 +1,446 @@
+        key = indexer.tuple
+        if len(key) != 1:
+            raise ValueError(
+                f"indexer with {len(key)} cannot be used for variable with 1 dimension"
+            )
+        index = key[0]
+        if isinstance(index, integer_types):
+            return self._converter.to_coordinate(index)
+        else:
+            return self._converter.to_coordinates(index)
+
+    @property
+    def dtype(self) -> np.dtype:
+        """Data type of the backend array."""
+        return self._converter.dtype
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        """Shape of the backend array."""
+        return self._converter.shape
+
+
+class TileDBDenseArrayWrapper(BackendArray):
+    """A backend array wrapper for a TileDB attribute.
+
+    This class is not intended to accessed directly. Instead it should be used
+    through a :class:`LazilyIndexedArray` object.
+    """
+
+    __slots__ = ("_array_args", "_index_converters", "shape", "dtype")
+
+    def __init__(self, attr, uri, key, timestamp, index_converters):
+        """
+        Parameters
+        ----------
+        attr : tiledb.Attr
+            The TileDB attribute being wrapped by this routine. Must be one of the
+            attributes in the array at the provided URI.
+        uri : str
+            Uniform Resoure Identifier (URI) for TileDB array. May be a path to a
+            local TileDB array or a URI for a remote resource.
+        key : Optional[str]
+            If not None, the key for accessing the TileDB array at the provided URI.
+        timestamp : Optional[int]
+            If not None, time in milliseconds to open the array at.
+        ctx : Optional[tiledb.Ctx]
+            If not None, a TileDB context manager object.
+        index_converters : Tuple[TileDBIndexConverter, ...]
+            The TileDBIndexConverters needed to convert each xarray index to its
+            corresponding TileDB coordinates.
+        """
+        if attr.isanon:
+            raise NotImplementedError(
+                "Support for anonymnous TileDB attributes has not been implemented."
+            )
+        self.dtype = attr.dtype
+        self._array_args = (uri, "r", key, timestamp, attr.name)
+        self._index_converters = index_converters
+        self.shape = tuple(converter.size for converter in index_converters)
+
+    def __getitem__(self, indexer: ExplicitIndexer):
+        xarray_indices = indexer.tuple
+        if len(xarray_indices) != len(self._index_converters):
+            raise ValueError(
+                f"key of length {len(xarray_indices)} cannot be used for a TileDB array"
+                f" of length {len(self._index_converters)}"
+            )
+        shape = tuple(
+            len(range(converter.size)[index] if isinstance(index, slice) else index)
+            for index, converter in zip(xarray_indices, self._index_converters)
+            if not isinstance(index, integer_types)
```

Comment:
Why exclude integer indexes here?

Reply:
In xarray, if a dimension has an integer look-up, then it collapses the dimension on that value. For example,

```
x = xr.DataArray(np.arange(16).reshape(4,4))
print(x[0, :])
```
will yield
```
<xarray.DataArray (dim_1: 4)>
array([0, 1, 2, 3])
Dimensions without coordinates: dim_1
```